File: serialize_deserialize/serialize_deserialize.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:

    // Serialize a tree to a string
    string serialize(TreeNode* root) {
        if (!root)
            return "null,";

        string serialized_tree = to_string(root->val) + ",";
        serialized_tree += serialize(root->left);
        serialized_tree += serialize(root->right);

        return serialized_tree;
    }

    // Deserialize a string to a tree
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        return deserialize_helper(ss);
    }

private:
    TreeNode* deserialize_helper(stringstream& ss) {
        string val;
        getline(ss, val, ',');
        if (val == "null")
            return nullptr;

        TreeNode* node = new TreeNode(stoi(val));
        node->left = deserialize_helper(ss);
        node->right = deserialize_helper(ss);

        return node;
    }
};

// Function to print the tree
void printTree(TreeNode* root) {
    if (root) {
        cout << root->val << " ";
        printTree(root->left);
        printTree(root->right);
    }
}

// Test
int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(4);
    root->right->right = new TreeNode(5);

    Codec codec;
    string serialized_tree = codec.serialize(root);
    cout << "Serialized tree: " << serialized_tree << endl;

    TreeNode* deserialized_root = codec.deserialize(serialized_tree);
    cout << "Deserialized tree: " << endl;
    printTree(deserialized_root); // In cây nhị phân đã deserialize
    return 0;
    
    return 0;
}

File: serialize_deserialize/serialize_deserialize.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root: TreeNode) -> str:
        def serialize_helper(node):
            if not node:
                return 'null,'
            serialized_tree = str(node.val) + ','
            serialized_tree += serialize_helper(node.left)
            serialized_tree += serialize_helper(node.right)
            return serialized_tree
        
        return serialize_helper(root)

    def deserialize(self, data: str) -> TreeNode:
        def deserialize_helper(nodes):
            val = next(nodes)
            if val == 'null':
                return None
            node = TreeNode(int(val))
            node.left = deserialize_helper(nodes)
            node.right = deserialize_helper(nodes)
            return node
        
        nodes = iter(data.split(','))
        return deserialize_helper(nodes)

# Function to print the tree
def printTree(root):
    if root:
        print(root.val, end=" ")
        printTree(root.left)
        printTree(root.right)

# Test
root = TreeNode(1)
root.left = TreeNode(2)
root.right = TreeNode(3)
root.right.left = TreeNode(4)
root.right.right = TreeNode(5)

codec = Codec()
serialized_tree = codec.serialize(root)
print("Serialized tree:", serialized_tree)

deserialized_root = codec.deserialize(serialized_tree)
print("Deserialized tree:")
printTree(deserialized_root) # In cây nhị phân đã deserialize


